text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+Using the dictionary above, we can compress to:
+
+```
+std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $6>>
+```
+
+A couple of things to note:
+
+ - The first occurrence of a dictionary entry is never substituted. We don't store the dictionary anywhere and need to be able to reconstruct it from the compressed version.
+ - Longer substitutions are preferred to shorter ones. `std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $4::IntoIter<u32>>>` would also decompress to the original version but the compiler is supposed to always pick the longest substitution available.
+
+The mangled version of a substitution marker is `S <key - 1> _` (and `S_` for key `0`) like in the Itanium mangling. So the above definition would be mangled to:
+
+```
+_RN12std_a1b2c3d44iter5ChainINS0_3ZipINS_3vec8IntoIterIjEES5_EEE
+```
+
+The uncompressed version would be:
+```
+_RN12std_a1b2c3d44iter5ChainIN12std_a1b2c3d44iter3ZipIN12std_a1b2c3d43vec8IntoIterIjEEN12std_a1b2c3d43vec8IntoIterIjEEEEE
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The reference-level explanation consists of three parts:
+
+1. A specification of the syntax mangled names conform to.
+2. A specification of the compression scheme.
+3. A mapping of Rust entities to the mangling syntax.
+
+For implementing a demangler, only the first two sections are of interest, that is, a demangler only needs to understand syntax and compression of names, but it does not have to care about how the compiler generates mangled names.
+
+
+## Syntax Of Mangled Names
+
+The syntax of mangled names is given in extended Backus-Naur form:
+
+ - Non-terminals are within angle brackets (as in `<name-prefix>`)
+ - Terminals are within quotes (as in `"_R"`),
+ - Optional parts are in brackets (as in `[<decimal>]`),
+ - Repetition (zero or more times) is signified by curly braces (as in `{ <name-prefix> }`)
+ - Comments are marked with `//`.
+
+Mangled names conform to the following grammar:
+
+```
+// The <decimal-number> specifies the encoding version.
+<symbol-name> = "_R" [<decimal-number>] <absolute-path> [<instantiating-crate>]
+
+<absolute-path> = "N" <path-prefix> [<generic-arguments>] "E"
```

Comment:
I guess these choices (`N`, `E`) and some of the basic types, happen to match the Itanium ABI?

I was thinking `P` might make more sense for paths, and `I` or `J` as a "closing bracket" (more so than `E`) but this is kind of a pointless bikeshed, since we don't expect people to read these themselves.

Reply:
Yeah, I guess `E` makes sense as an "end" marker. Without a strong reason, I'd just stick to `N` and `E`. They are as good as any.